### PR TITLE
get picks for aggregator status

### PR DESCRIPTION
**Pull request by deads2k**
xref https://github.com/openshift/origin/issues/14500

Aggregator status usage didn't come until long after the API was added, so this pulls in the most important picks that are precursors to it so that the conflicts aren't crazy.  The picks came in relatively clean.  I've kept the fitting commits separate for ease.

@liggitt

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
Alright, the last commit is a change that is needed to wire openshift into a more "normal" aggregated API server shape so that our admission chain works after the aggregator picks we needed from upstream.  It is mostly a shuffle, but it is fairly large.  Luckily, our integration test flows cover this.  More refactoring is necessary, but this is about as small as I could get and have it work.

@soltysh @mfojtik @sttts ptal.
@liggitt @eparis this is a structural change that has almost no shot of making it through the origin queue in a reasonable time due to likely conflict/rebase cycles that push me to the back.  I'm requesting a green button on lgtm and green tests.

**Comment by deads2k** [comment]
> @liggitt @eparis this is a structural change that has almost no shot of making it through the origin queue in a reasonable time due to likely conflict/rebase cycles that push me to the back. I'm requesting a green button on lgtm and green tests.

for perspective, the queue is currently 26 deep and working on a pull that was tagged for merge on Monday.

File: pkg/cmd/server/origin/apiserver.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
+package origin
```
Comment (sttts) [inline pkg/cmd/server/origin/apiserver.go]: openshift-apiserver.go ?

Comment (sttts) [inline pkg/cmd/server/origin/apiserver.go]: or originapi-apiserver.go

File: pkg/cmd/server/origin/nonapiserver.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+	// these are only needed for the template service broker, which should move out
+	KubeClientInternal          kclientsetinternal.Interface
+	EnableTemplateServiceBroker bool
+	TemplateInformers           templateinformer.SharedInformerFactory
+	// c.Options.TemplateServiceBrokerConfig.TemplateNamespaces
+	TemplateNamespaces []string
+}
+
+// OpenshiftNonAPIServer is only responsible for serving the APIs for Openshift
```
Comment (sttts) [inline pkg/cmd/server/origin/nonapiserver.go]: outdated comment?

File: pkg/cmd/server/origin/master.go [hunk 3]
```diff
@@ -2,214 +2,145 @@ package origin
...
-	apiserver := kmaster.GenericAPIServer.PrepareRun()
+	kubeAPIServer, err := kubeAPIServerConfig.Complete().New(openshiftNonAPIServer.GenericAPIServer)
+	if err != nil {
+		glog.Fatalf("Failed to launch master: %v", err)
+	}
+	// TODO this goes away in 3.7 after we commit to the aggregator always being on (even if its just in local mode).
+	// this is installing the openshift APIs into the kubeapiserver
+	installAPIs(openshiftAPIServerConfig, kubeAPIServer.GenericAPIServer)
+	// this sets up the openapi endpoints
+	preparedKubeAPIServer := kubeAPIServer.GenericAPIServer.PrepareRun()
```
Comment (sttts) [inline pkg/cmd/server/origin/master.go]: don't like that we call `PrepareRun()` only here and not for the others. Is it the same in kube-apiserver?

Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: > don't like that we call PrepareRun() only here and not for the others. Is it the same in kube-apiserver?

Yeah.  openapi is the source of many sins

File: pkg/cmd/server/origin/apiserver.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	for group, versions := range groupVersions {
+		apiGroupInfo := genericapiserver.NewDefaultAPIGroupInfo(group, kapi.Registry, kapi.Scheme, kapi.ParameterCodec, kapi.Codecs)
+
+		for _, version := range versions {
+			gv := schema.GroupVersion{Group: group, Version: version}
+			apiGroupInfo.VersionedResourcesStorageMap[version] = storage[gv]
+
+			// TODO all of our groups currently have one version, by the time we get more than one, these should be split up
+			// into their own api servers
+			apiGroupInfo.GroupMeta.GroupVersion = gv
```
Comment (sttts) [inline pkg/cmd/server/origin/apiserver.go]: Why not keep isPreferredGroupVersion?

File: pkg/cmd/server/origin/apiserver.go [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+			glog.Infof("Starting Origin API at %%s%s/%s/%s", api.GroupPrefix, gv.Group, gv.Version)
```
Comment (sttts) [inline pkg/cmd/server/origin/apiserver.go]: note the %%s. This is substituted somewhere else in the old code.

Comment (sttts) [inline pkg/cmd/server/origin/apiserver.go]: and all the other places with `%%s`

File: pkg/cmd/server/origin/nonapiserver.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	return completedOpenshiftNonAPIConfig{c}
+}
+
+// SkipComplete provides a way to construct a server instance without config completion.
+func (c *OpenshiftNonAPIConfig) SkipComplete() completedOpenshiftNonAPIConfig {
+	return completedOpenshiftNonAPIConfig{c}
+}
+
+func (c completedOpenshiftNonAPIConfig) New(delegationTarget genericapiserver.DelegationTarget, stopCh <-chan struct{}) (*OpenshiftNonAPIServer, error) {
+	genericServer, err := c.OpenshiftNonAPIConfig.GenericConfig.SkipComplete().New("openshift-nonapiserver", delegationTarget) // completion is done in Complete, no need for a second time
```
Comment (sttts) [inline pkg/cmd/server/origin/nonapiserver.go]: nonapiserver is strange somehow. openshift-non-api-routes?

Comment (deads2k) [inline pkg/cmd/server/origin/nonapiserver.go]: > nonapiserver is strange somehow. openshift-non-api-routes?

Not quite as much fun, but sure.

File: pkg/cmd/server/origin/storage.go [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+// per API group to give us legacy and current storage
+func (c OpenshiftAPIConfig) GetRestStorage() (map[schema.GroupVersion]map[string]rest.Storage, error) {
+	// TODO sort out who is using this and why.  it was hardcoded before the migration and I suspect that it is being used
+	// to serialize out objects into annotations.
+	externalVersionCodec := kapi.Codecs.LegacyCodec(schema.GroupVersion{Group: "", Version: "v1"})
+
+	//TODO/REBASE use something other than c.KubeClientsetInternal
+	nodeConnectionInfoGetter, err := kubeletclient.NewNodeConnectionInfoGetter(c.KubeClientExternal.CoreV1().Nodes(), *c.KubeletClientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to configure the node connection info getter: %v", err)
```
Comment (sttts) [inline pkg/cmd/server/origin/storage.go]: nit: lower case

Comment (sttts) [inline pkg/cmd/server/origin/storage.go]: everywhere below as well

File: pkg/cmd/server/origin/storage.go [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+		TLSClientConfig: restclient.TLSClientConfig{
+			Insecure: true,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to configure a default transport for importing: %v", err)
+	}
+
+	buildStorage, buildDetailsStorage, err := buildetcd.NewREST(c.GenericConfig.RESTOptionsGetter)
+	if err != nil {
```
Comment (sttts) [inline pkg/cmd/server/origin/storage.go]: would write those if-clause in one line, if gofmt accepts that. Func gets very long with 50% error  handling.

Comment (deads2k) [inline pkg/cmd/server/origin/storage.go]: > would write those if-clause in one line, if gofmt accepts that. Func gets very long with 50% error handling.

I miss java too :)

This function is on the chopping block.  I wouldn't get overly caught up in it.

File: pkg/cmd/server/origin/master.go [hunk 9]
```diff
@@ -419,612 +318,6 @@ func (c *MasterConfig) InitializeObjects() {
...
-			w.Write([]byte("ok"))
-
-		} else {
-			w.WriteHeader(http.StatusServiceUnavailable)
-		}
-	})
-}
-
-// initMetricsRoute initializes an HTTP endpoint for metrics.
-func initMetricsRoute(apiContainer *restful.Container, path string) {
```
Comment (sttts) [inline pkg/cmd/server/origin/master.go]: this was dead code?

File: pkg/cmd/server/origin/master.go [hunk 10]
```diff
@@ -419,612 +318,6 @@ func (c *MasterConfig) InitializeObjects() {
...
-		Scheme: kapi.Scheme,
-		// version.ParameterCodec = runtime.NewParameterCodec(kapi.Scheme)
-		ParameterCodec:              kapi.ParameterCodec,
-		NegotiatedSerializer:        kapi.Codecs,
-		SubresourceGroupVersionKind: map[string]schema.GroupVersionKind{},
-	}
-
-	storage := make(map[string]rest.Storage)
-	for k, v := range all {
-		if excludedV1Types.Has(k) {
```
Comment (sttts) [inline pkg/cmd/server/origin/master.go]: this was empty I guess.

File: pkg/cmd/server/origin/nonapiserver.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	GenericConfig *genericapiserver.Config
+
+	// these are only needed for the controller endpoint which should be moved out and made an optional
+	// add-on in the chain (as the final delegate) when running an all-in-one
+	EnableControllers bool
+	ControllerPlug    plug.Plug
+
+	// c.Options.OAuthConfig.MasterPublicURL
+	MasterPublicURL string
+	// c.Options.OAuthConfig != nil
```
Comment (sttts) [inline pkg/cmd/server/origin/nonapiserver.go]: Better comment? These code comments will get outdated.

File: pkg/cmd/server/origin/storage.go [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+	"github.com/openshift/origin/pkg/security/registry/podsecuritypolicyreview"
+	"github.com/openshift/origin/pkg/security/registry/podsecuritypolicyselfsubjectreview"
+	"github.com/openshift/origin/pkg/security/registry/podsecuritypolicysubjectreview"
+	oscc "github.com/openshift/origin/pkg/security/scc"
+
+	// register api groups
+	_ "github.com/openshift/origin/pkg/api/install"
+)
+
+// TODO this function needs to be broken apart with each API group owning their own storage, probably with two method
```
Comment (sttts) [inline pkg/cmd/server/origin/storage.go]: yes, please. 300 lines.

File: pkg/cmd/server/origin/master.go (hunk 9 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: yes

File: pkg/cmd/server/origin/master.go (hunk 10 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: yes

File: pkg/cmd/server/origin/storage.go (hunk 12 above)
Comment (deads2k) [inline pkg/cmd/server/origin/storage.go]: > yes, please. 300 lines.

follow up

Comment (sttts) [inline pkg/cmd/server/origin/storage.go]: fine with me.

File: pkg/cmd/server/origin/storage.go (hunk 8 above)
Comment (sttts) [inline pkg/cmd/server/origin/storage.go]: True.

File: pkg/cmd/server/origin/apiserver.go (hunk 4 above)
Comment (deads2k) [inline pkg/cmd/server/origin/apiserver.go]: restored

File: pkg/cmd/server/origin/nonapiserver.go (hunk 11 above)
Comment (deads2k) [inline pkg/cmd/server/origin/nonapiserver.go]: they were breadcrumbs for me. removing

**Comment by deads2k** [comment]
Comments addressed and some more tests fixed.  So glad we have tests.

**Comment by sttts** [comment]
Looks good from my side. Having some more eyes on this wouldn't harm though. Quite a bit of code moved and split apart.

**Comment by deads2k** [comment]
Appears to be clean of conflicts with the next two in the queue.  Plan to merge on green.

**Comment by deads2k** [comment]
@liggitt @sttts Hit a snag with admission.  See the last two commits for a "fix" that works until we aggregate in 3.7.

**Comment by deads2k** [comment]
This breaks the back of the problem, but then we have at least one TODO that needs doing and the generated code problem.  I'd like to see this merged to get those a lot smaller.

**Comment by deads2k** [comment]
Re[test]

File: pkg/cmd/server/origin/master.go [hunk 13]
```diff
@@ -128,7 +128,14 @@ func (c *MasterConfig) Run(kubeAPIServerConfig *kubeapiserver.Config, assetConfi
 	}
 	// TODO this goes away in 3.7 after we commit to the aggregator always being on (even if its just in local mode).
 	// this is installing the openshift APIs into the kubeapiserver
+	// ok, this is a big side-effect.  Openshift APIs run a different admission chain (always have), but since
+	// we're going through a "normal" API installation in the wrong server, we need to switch the admission chain
+	// *only while we're installing these APIs*.  There are tests that make sure this works and doesn't drop
+	// plugins and we'll remove it once we're aggregating
+	kubeAPIServer.GenericAPIServer.SetAdmission(openshiftAPIServerConfig.GenericConfig.AdmissionControl)
```
Comment (sttts) [inline pkg/cmd/server/origin/master.go]: ugly, but ok.

Wondering which of these temporary hacks we can move into the if-clause below for the non-aggregating run mode?

**Comment by sttts** [comment]
@deads2k wouldn't call it peace, but I am ok with those two (temporary) commits.

File: pkg/cmd/server/origin/master.go (hunk 13 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: > Wondering which of these temporary hacks we can move into the if-clause below for the non-aggregating run mode?

Can't.  These are always needed to get the admission chain right until we're always aggregating. 
 :(

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
flaked on https://github.com/openshift/origin/issues/14573

**Comment by deads2k** [comment]
Green except for the flake.  Blocks aggregator usage. Merging as planned.

**Pull request merged**